### PR TITLE
Feature/donor receiver roles

### LINKED ISSUE #111: Reputation receiver and donator role
allow to define roles which indicate whether a user can
- receive reputation
- give reputation

This will be turned off until a role is added.

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Add roles which are required to give and receive reputation.
When no roles are set everyone can give and receive reputation.

Close #111

### CODE REVIEW

File: src/main/resources/locale.properties [hunk 1]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputation of %NAME%
 command.roles.description=Manage reputation roles.
 command.roles.sub.add=Add a reputation role.
 command.roles.sub.add.added=Added role %ROLE% with reputation %POINTS%.
+command.roles.sub.addDonor=Add a donor role which can give reputation
+command.roles.sub.addDonor.add=Added donor role %ROLE%
+command.roles.sub.addReceiver=Add a receiver role, which can receive reputation
+command.roles.sub.addReceiver.add=Added receiver role %role%
```
Comment (Taucher2003) [inline src/main/resources/locale.properties:111]:
```suggestion
command.roles.sub.addReceiver.add=Added receiver role %ROLE%
```

File: src/main/resources/locale_en_US.properties [hunk 2]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputation of %NAME%
 command.roles.description=Manage reputation roles.
 command.roles.sub.add=Add a reputation role.
 command.roles.sub.add.added=Added role %ROLE% with reputation %POINTS%.
+command.roles.sub.addDonor=Add a donor role which can give reputation
+command.roles.sub.addDonor.add=Added donor role %ROLE%
+command.roles.sub.addReceiver=Add a receiver role, which can receive reputation
+command.roles.sub.addReceiver.add=Added receiver role %role%
```
Comment (Taucher2003) [inline src/main/resources/locale_en_US.properties:111]:
```suggestion
command.roles.sub.addReceiver.add=Added receiver role %ROLE%
```

File: src/main/resources/locale_es_ES.properties [hunk 3]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputación de %NAME%
 command.roles.description=Gestionar los roles de reputación.
 command.roles.sub.add=Añadir un rol de reputación.
 command.roles.sub.add.added=Añadir rol %ROLE% con reputación %POINTS%.
+command.roles.sub.addDonor=Añadir un rol de donante que pueda dar reputación
+command.roles.sub.addDonor.add=Se ha añadido el rol de donante %ROLE%.
+command.roles.sub.addReceiver=Añadir un rol de receptor, que puede recibir reputación
+command.roles.sub.addReceiver.add=Añadido el rol de receptor %role%.
```
Comment (Taucher2003) [inline src/main/resources/locale_es_ES.properties:111]:
```suggestion
command.roles.sub.addReceiver.add=Añadido el rol de receptor %ROLE%.
```

File: src/main/resources/locale_fr_FR.properties [hunk 4]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Réputation de %NAME%
 command.roles.description=Gère les rôles de réputation.
 command.roles.sub.add=Ajouter un rôle de réputation.
 command.roles.sub.add.added=Ajouté le rôle %ROLE% avec la réputation %POINTS%.
+command.roles.sub.addDonor=Ajouter un rôle de donateur qui peut donner une réputation
+command.roles.sub.addDonor.add=Ajout du rôle de donneur %ROLE%.
+command.roles.sub.addReceiver=Ajouter un rôle de récepteur, qui peut recevoir la réputation
+command.roles.sub.addReceiver.add=Ajout du rôle de récepteur %role%
```
Comment (Taucher2003) [inline src/main/resources/locale_fr_FR.properties:111]:
```suggestion
command.roles.sub.addReceiver.add=Ajout du rôle de récepteur %ROLE%
```

File: src/main/resources/locale_pt_PT.properties [hunk 5]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputação de %NAME%
 command.roles.description=Gerir papéis de reputação.
 command.roles.sub.add=Adicionar um papel de reputação.
 command.roles.sub.add.added=Added role %ROLE% com reputação %POINTS%.
+command.roles.sub.addDonor=Acrescentar um papel de doador que possa dar reputação
+command.roles.sub.addDonor.add=Papel do doador acrescentado %ROLE%ROLE
```
Comment (Taucher2003) [inline src/main/resources/locale_pt_PT.properties:109]:
```suggestion
command.roles.sub.addDonor.add=Papel do doador acrescentado %ROLE%
```

File: src/main/resources/locale_pt_PT.properties [hunk 6]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputação de %NAME%
...
+command.roles.sub.addReceiver=Acrescentar um papel de receptor, que pode receber reputação
+command.roles.sub.addReceiver.add=Adicionada a função de receptor %role%role
```
Comment (Taucher2003) [inline src/main/resources/locale_pt_PT.properties:111]:
```suggestion
command.roles.sub.addReceiver.add=Adicionada a função de receptor %ROLE%
```

**Review by Taucher2003 — CHANGES_REQUESTED** [review]

**Review by Taucher2003 — APPROVED** [review]

File: src/main/java/de/chojo/repbot/commands/Roles.java [hunk 7]
```diff
@@ -51,24 +73,50 @@ public Roles(DataSource dataSource, ILocalizer loc) {
...
-        }
-
-        if ("remove".equalsIgnoreCase(subCmd)) {
-            return remove(eventWrapper, context.subContext(subCmd));
-        }
         if ("managerRole".equalsIgnoreCase(subCmd)) {
             return managerRole(eventWrapper, context.subContext(subCmd));
         }
+
+        if(StringUtil.contains(subCmd, "add", "remove", "addDonor", "addReceiver", "removeDonor", "removeReceiver")){
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/commands/Roles.java:86]:
```suggestion
        if (StringUtil.contains(subCmd, "add", "remove", "addDonor", "addReceiver", "removeDonor", "removeReceiver")) {
```

File: src/main/java/de/chojo/repbot/commands/Roles.java [hunk 8]
```diff
@@ -40,6 +46,22 @@ public Roles(DataSource dataSource, ILocalizer loc) {
...
+                                .build()
+                        )
+                        .add("removedonor", "command.roles.sub.removeDonor", argsBuilder()
+                                .add(OptionType.ROLE, "role", "role", true)
+                                .build()
+                        )
+                        .add("removereceiver", "command.roles.sub.removeReceiver", argsBuilder()
+                                .add(OptionType.ROLE, "role", "role", true)
+                                .build()
+                        )
```
Comment (dlsf) [inline src/main/java/de/chojo/repbot/commands/Roles.java:64]: I'd suggest the following for better readability, if it's possible:
```suggestion
                        .add("addDonor", "command.roles.sub.addDonor", argsBuilder()
                                .add(OptionType.ROLE, "role", "role", true)
                                .build()
                        )
                        .add("addReceiver", "command.roles.sub.addReceiver", argsBuilder()
                                .add(OptionType.ROLE, "role", "role", true)
                                .build()
                        )
                        .add("removeDonor", "command.roles.sub.removeDonor", argsBuilder()
                                .add(OptionType.ROLE, "role", "role", true)
                                .build()
                        )
                        .add("removeReceiver", "command.roles.sub.removeReceiver", argsBuilder()
                                .add(OptionType.ROLE, "role", "role", true)
                                .build()
                        )
```

File: src/main/resources/locale_de_DE.properties [hunk 9]
```diff
@@ -113,6 +117,10 @@ command.roles.sub.managerRole.set=Die Manager-Rolle wurde auf %ROLE% gesetzt.
 command.roles.sub.remove=Entfernt eine Reputationsrolle.
 command.roles.sub.remove.notARepRole=Diese Rolle ist keine Reputationsrolle.
 command.roles.sub.remove.removed=Die Rolle %ROLE% wurde entfernt.
+command.roles.sub.removeDonor=Entferne eine Spender Rolle
+command.roles.sub.removeDonor.remove=Spender Rolle %ROLE% entfernt
+command.roles.sub.removeReceiver=Entferne eine Empfänger Rolle
+command.roles.sub.removeReceiver.remove=Empfänger Rolle %ROLE% entfernt
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:123]:
```suggestion
command.roles.sub.removeDonor=Entfernt eine Spender-Rolle.
command.roles.sub.removeDonor.remove=Spender-Rolle %ROLE% entfernt.
command.roles.sub.removeReceiver=Entfernt eine Empfänger-Rolle.
command.roles.sub.removeReceiver.remove=Empfänger-Rolle %ROLE% entfernt.
```

File: src/main/resources/locale_de_DE.properties [hunk 10]
```diff
@@ -105,6 +105,10 @@ command.reputation.profile.title=Reputation von %NAME%
 command.roles.description=Verwaltet die Reputationsrollen.
 command.roles.sub.add=Fügt eine Reputationsrolle hinzu.
 command.roles.sub.add.added=Die Rolle %ROLE% mit der benötigten Reputation %POINTS% wurde erfolgreich hinzugefügt.
+command.roles.sub.addDonor=Füge eine Spenderrolle hinzu, welche Reputation erhalten kann
+command.roles.sub.addDonor.add=Spender Rolle %ROLE% hinzugefügt
+command.roles.sub.addReceiver=Füge eine Empfängerrolle hinzu, welche Reputation vergeben kann
+command.roles.sub.addReceiver.add=Empfänger Rolle %ROLE% hinzugefügt
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:111]: `Empfänger-Rolle` sounds a little bit odd but I'm not sure if there is a better alternative. 
```suggestion
command.roles.sub.addDonor=Fügt eine Spender-Rolle hinzu, welche Reputation erhalten kann.
command.roles.sub.addDonor.add=Spender-Rolle %ROLE% hinzugefügt.
command.roles.sub.addReceiver=Füge eine Empfänger-Rolle hinzu, welche Reputation vergeben kann.
command.roles.sub.addReceiver.add=Empfänger-Rolle %ROLE% hinzugefügt.
```

**Review by dlsf — CHANGES_REQUESTED** [review]

File: src/main/java/de/chojo/repbot/commands/Roles.java (hunk 8 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/commands/Roles.java:64]: Subcommand have to be all lowercase. otherwise the slash commands will break.

**Review by dlsf — APPROVED** [review]

**Pull request merged**